test/core/generate_mipmaps_sampler2d.cpp

```diff
@@ -7,40 +13,60 @@
 namespace generate_mipmaps
 {
 	template <typename genType>
-	int test(gli::format Format, genType const & Black, genType const & Color, std::size_t Size, gli::filter Filter)
+	int test(gli::format Format, const genType& BaseColor, const genType& TargetColor, std::size_t Size, gli::filter Filter)
 	{
 		int Error = 0;
 
-		gli::texture2d Texture(Format, gli::texture2d::extent_type(static_cast<gli::texture2d::extent_type::value_type>(Size)));
-		Texture.clear(Black);
-		Texture[0].clear(Color);
+		// create 2-D texture of appropriate size and clear with texels of given colors
+		gli::texture2d Texture(Format,
+			static_cast<gli::texture2d::extent_type>(static_cast<gli::texture2d::extent_type::value_type>(Size)));
+		Texture.clear(BaseColor);
+		Texture[0].clear(TargetColor);
+
+		// get texel at position #0 from top mipmap level
+		const genType TopTexel = Texture.load<genType>(static_cast<gli::texture2d::extent_type>(0),
+			Texture.max_level());
 
-		genType const LoadC = Texture.load<genType>(gli::texture2d::extent_type(0), Texture.max_level());
-		if(Texture.levels() > 1)
-			Error += LoadC == Black ? 0 : 1;
+		// If texture has multiple mipmap levels, texel from top level must match base color
+		if (Texture.levels() > 1)
+			Error += (TopTexel == BaseColor) ? 0 : 1;
 
+		// create a texture view and a sampler
 		gli::texture2d TextureView(gli::view(Texture, 0, 0));
 		gli::fsampler2D SamplerA(gli::texture2d(gli::duplicate(Texture)), gli::WRAP_CLAMP_TO_EDGE);
-		SamplerA.generate_mipmaps(gli::FILTER_LINEAR);
```

Comment:
`gli::FILTER_LINEAR` fails some of the test iterations. It was replaced with the value of `Filter` passed to the `test` function.